generate-rs/src/parser.rs

```diff
@@ -0,0 +1,593 @@
+fn optionalize(name: &str, optional: bool) -> String {
+    unimplemented!()
+    //return 'Option<{}>'.format(name) if optional else name
+}
+
+
+fn replace_keywords(name: &str) -> String {
+    unimplemented!()
+    // return {
+    //     'type': 'typ',
+    //     'enum': 'enums',
+    // }.get(name, name)
+}
+
+fn capitalize_first(name: &str) -> String {
+    unimplemented!()
+    // if len(name) == 0:
+    //     return name
+    // return name[0].upper() + name[1:]
+}
+
+fn rust_identifier(name: &str) -> String {
+    unimplemented!()
+    // def sanitize(s):
+    //     return s.replace('$', 'dollar').replace('#', 'hash').replace('.', '_')
+
+    // def r(c):
+    //     if not c.isupper():
+    //         return c
+    //     return '_' + c.lower()
+
+    // return ''.join([(r(c) if i > 0 else c.lower()) for i, c in enumerate(sanitize(name))])
+}
+
+
+fn snake_to_camel(name: &str) -> String {
+    unimplemented!()
+    // dest = []
+    // capitalize = True
+    // for c in name:
+    //     if c == '_':
+    //         capitalize = True
+    //         continue
+    //     if capitalize:
+    //         dest.append(c.upper())
+    //         capitalize = False
+    //         continue
+    //     dest.append(c)
+    // return ''.join(dest)
+}
+
+fn global_params_name(api_name: &str) -> String {
+    unimplemented!()
+    // return snake_to_camel(api_name + 'Params')
+}
+
+
+fn parse_schema_types(name: &str, schema: &str, optional: bool, parents: Vec<String>) {
```

Comment:
this can probably be tightened and split up into subroutines a bit, and compared to any official standards (if existing)